u_fingerprinter_test.go

```diff
@@ -0,0 +1,408 @@
+				// We can't really compare padding extensions in this way
+				continue
+			}
+
+			generatedExtension := generatedUConn.Extensions[i]
+			checkUTLSExtensionsEquality(t, originalExtension, generatedExtension)
+		}
+	}
+
+	fieldsToTest := []string{
+		"Vers", "CipherSuites", "CompressionMethods", "NextProtoNeg", "ServerName", "OcspStapling", "Scts", "SupportedCurves",
+		"SupportedPoints", "TicketSupported", "SupportedSignatureAlgorithms", "SecureRenegotiation", "SecureRenegotiationSupported", "AlpnProtocols",
+		"SupportedSignatureAlgorithmsCert", "SupportedVersions", "KeyShares", "EarlyData", "PskModes", "PskIdentities", "PskBinders",
+	}
+
+	for _, field := range fieldsToTest {
+		compareClientHelloFields(t, field, uconn.HandshakeState.Hello, generatedUConn.HandshakeState.Hello)
+	}
+}
+
+func TestUTLSFingerprintClientHello(t *testing.T) {
+	clientHellosToTest := []ClientHelloID{
+		HelloChrome_58, HelloChrome_70, HelloChrome_83, HelloFirefox_55, HelloFirefox_63, HelloIOS_11_1, HelloIOS_12_1, HelloRandomized, HelloRandomizedALPN, HelloRandomizedNoALPN}
+
+	serverNames := []string{"foobar"}
+
+	for _, clientHello := range clientHellosToTest {
+		for _, serverName := range serverNames {
+			t.Logf("checking fingerprint generated client hello spec against %v and server name: %v", clientHello, serverName)
+			checkUTLSFingerPrintClientHello(t, clientHello, "foobar")
+		}
+	}
+}
+
+func TestUTLSFingerprintClientHelloBluntMimicry(t *testing.T) {
+	serverName := "foobar"
+	var extensionId uint16 = 0xfeed
+	extensionData := []byte("random data")
+
+	specWithGeneric, err := utlsIdToSpec(HelloChrome_Auto)
+	if err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+	specWithGeneric.Extensions = append(specWithGeneric.Extensions, &GenericExtension{extensionId, extensionData})
+
+	uconn := UClient(&net.TCPConn{}, &Config{ServerName: serverName}, HelloCustom)
+
+	if err := uconn.ApplyPreset(&specWithGeneric); err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+	if err := uconn.BuildHandshakeState(); err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+
+	f := &Fingerprinter{}
+	_, err = f.FingerprintClientHello(uconn.HandshakeState.Hello.Raw)
+	if err == nil {
+		t.Errorf("expected error generating spec from client hello with GenericExtension")
+	}
+
+	f = &Fingerprinter{AllowBluntMimicry: true}
+	generatedSpec, err := f.FingerprintClientHello(uconn.HandshakeState.Hello.Raw)
+	if err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+
+	for _, ext := range generatedSpec.Extensions {
+		if genericExtension, ok := (ext).(*GenericExtension); ok {
+			if genericExtension.Id == extensionId && bytes.Equal(genericExtension.Data, extensionData) {
+				return
+			}
+		}
+	}
+	t.Errorf("generated ClientHelloSpec with BluntMimicry did not correctly carry over generic extension")
+}
+
+func TestUTLSFingerprintClientHelloKeepPSK(t *testing.T) {
+	// Handshake Protocol: Client Hello
```

Comment:
I did this in some `u_conn.go` tests as well. Do you think this is a reasonable format for using a raw/captured Client Hello?

Reply:
Seems reasonable